### ISSUE TITLE
Interactive input should works with -File parameter

### ISSUE THREAD

**Issue by iSazonov** [PowerShell/PowerShell#9494]
Come form #9362
The comment in code says that interactive input should works
- with -File:
https://github.com/PowerShell/PowerShell/blob/ceed73d7375cb98a199ceb19796bb2311b0b0002/src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs#L1133-L1141

- with -Command:
https://github.com/PowerShell/PowerShell/blob/ceed73d7375cb98a199ceb19796bb2311b0b0002/src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs#L1296-L1304

The comments differ in "no prompts" vs "with prompts" - it is not clear is it comments and behaviors right.

Also using dash is documented only for `-Command` (`not for -File`)
https://docs.microsoft.com/en-us/powershell/scripting/components/console/powershell.exe-command-line-help?view=powershell-6#-command

  
# Steps to reproduce

```powershell
pwsh -file C:\temp\q1.ps1 -
```

# Expected behavior

```none
pwsh -file C:\1\q1.ps1 -
<wait interactive inout>
```

# Actual behavior

```none
pwsh -file C:\1\q1.ps1 -
-
```

# Expected behavior

```none
pwsh -command -
<wait interactive inout>
```

# Actual behavior

```none
'-' was specified as the argument to -Command but standard input has not been redirected for this process.

Usage: pwsh[.exe] [[-File] <filePath> [args]]
                  [-Command { - | <script-block> [-args <arg-array>]
                                | <string> [<CommandParameters>] } ]
                  [-ConfigurationName <string>] [-CustomPipeName <string>]
                  [-EncodedCommand <Base64EncodedCommand>]
                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]
                  [-Interactive] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]
                  [-OutputFormat {Text | XML}] [-Version] [-WindowStyle <style>]
                  [-WorkingDirectory <directoryPath>]

       pwsh[.exe] -h | -Help | -? | /?

PowerShell Online Help https://aka.ms/pscore6-docs

All parameters are case-insensitive.
-
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.2.0-rc.1
PSEdition                      Core
GitCommitId                    6.2.0-rc.1-204-g262d86c691519ff7d5adacbae6424e2c87accce6
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iSazonov** [comment]
Perhaps we could discuss this in #3223.

**Comment by mklement0** [comment]
The `-` in `pwsh -file C:\temp\q1.ps1 -` is just a positional argument; `-` is only special as the _first_ argument after `-file` or `-command`; that is, if it is specified _instead of_ a file name / code snippet.

Do note that the primary use of these idioms is to provide input via _stdin_  / the pipeline, not _interactive_ use.

Given that, when called from the outside `-` should represent _stdin_, the following commands should all be equivalent on Unix:

```powershell
pwsh -noprofile -file -
pwsh -noprofile -file /dev/stdin
```

`pwsh -noprofile  -file -` is broken (as is `pwsh -noprofile -command -`, as demonstrated in your OP).

`pwsh -noprofile -file /dev/stdin` works in principle, but, once the input is submitted with ^D, it then displays pseudo-interactive _output_ behavior, as discussed in #3223.

(As an aside, `bash` actually prompts _line by line_ in this scenario,  because, Iin the Unix world, _non-redirected_, i.e., interactive stdin input is processed line by line, whereas _redirected_ stdin is typically processed as a whole. Again, however: interactive input is _not_ the typical or useful use case - if you want to submit commands interactively, one by one, just start a new session with `pwsh`.)

Returning to the typical use case:

The following two commands should be equivalent (with the exception of what the script reports in `$PSCommandPath`, `$PSScriptRoot`, ...):

* `pwsh -noprofile -file foo.ps1`
* `cat foo.ps1 | pwsh -noprofile -file -`

It is what enables the powerful `curl ... | sh` install-from-the-web idiom in POSIX-like shells - that is, feed the _content_ of a script to the shell rather than needing to provide a _file_['s path].

If `-file -` worked that way, there would really be no need for `-command -` at all.
Given that we cannot remove support for it, it should be made to work the same, but [soft]-deprecated.

There's also no reason to disallow _arguments_ to follow `-file -` - it's an unnecessary restriction that violates the equivalence of the two commands above. It prevents passing the script content via stdin / the pipeline while also accepting _arguments_.

(Given how `-command` currently works, there _is_ a good reason to disallow following `-command -` with arguments (the way that `-command` currently works - stitching together a script from _all_ arguments following `-command` - is highly unfortunate, but cannot be changed), but if we fix `-file -`, that restriction can then easily be bypassed by using `-file -` instead).

Note that it still makes sense to combine `-command` - _not_ with `-` - with stdin input: The `-command` argument then specifies the command to run, as usual, and stdin input becomes its _pipeline input_.

Here's a `bash` example:

```sh
$ printf 'hi\nthere\n' | bash -c 'cat -n'
     1	hi
     2	there
```

Regrettably, stdin input passed to `pwsh` currently does _not_ become pipeline input - see #9497 -  so that the following does _not_ work analogously:

```powershell
$ printf 'hi\nthere\n' | pwsh -noprofile -command '% { $i=0 } { "{0}`t{1}" -f ++$i, $_ }'
1   # !! stdin input did NOT become pipeline input
```

As a doleful aside: if you tried to run this _from PowerShell_, the broken quoting for external programs (PowerShell itself, in this case), would make the command _fail_.

**Comment by iSazonov** [comment]
> Do note that the primary use of these idioms is to provide input via stdin / the pipeline, not interactive use.

Documentation is silent about it that confuses me. Make sense to open new issue in Docs repo?

The question remains about the input from the console. Although it could also be documented that it is not supported.

**Comment by mklement0** [comment]
The docs definitely need improvement: The CLI help doesn't mention `-File -` at all, and doesn't mention the requirement of _redirected_ stdin input for `-Command -`

The current behavior is conceptually backwards and fundamentally broken, as shown in #3223:

* It is designed around _line-by-line_ processing, so clearly focused on _interactive_ use.

  * By contrast, the real value comes from providing commands via _redirected_ stdin  / piping.

* It is `-Command -` that is implemented with _redirected_ stdin input in mind - yet line-by-line process is _still_ applied (but at least the prompt string isn't printed the way it is with `-File -`), unlike when you pass a multi-line command string as an _argument_.

  * Conceptually, `-File -` makes more sense  for redirected stdin input - and it is the _only_ form needed.

> The question remains about the input from the console.

`pwsh -File -` without redirected stdin input simply starts a regular session - the same as not specifying `-File -` at all.
While technically somewhat defensible, it's certainly surprising.

There _is_ a way in which `-File -` could be useful with _interactive_ input:
If it allowed you to type an entire script ad hoc, and then submit it at once:

```powershell
$ pwsh -File -  # type the ad hoc script on the following lines, then submit with ^D (+ Enter on Windows)
get-date
get-childItem -Directory
^D
```

This is what `ruby` supports, for instance.

The caveat is that this form of supplying multi-line stdin input is probably more familiar to Unix users than Windows users, and it would be a first in the context of PowerShell.
(I suppose submitting an empty extra line could be taken as the end-of-input signal, the same way as with array-valued mandatory parameter prompts, but that preclude use of blank lines for readability.)

**Comment by kmccl922** [comment]
Stop sending me they damn messages taken my email address for more important information i get on this email account Thank You Please Stop

Sent from Yahoo Mail on Android 
 
  On Wed, Aug 25, 2021 at 12:29 AM, ***@***.***> wrote:   


Closed #9494.
